### ISSUE TITLE
Feature request: for fluentbit plugins for s3 too

### ISSUE THREAD

**Issue by rverma-jm** [aws/aws-for-fluent-bit#31]
Equivalent fluentd plugin: https://github.com/fluent/fluent-plugin-s3

Support both s3-sqs input and s3 outut.
Would be amazing if it can support writing in parquet.

**Comment by PettitWesley** [comment]
Hey @rverma-jm see the discussion in the linked core fluent/fluent-bit issue.

I'm curious about the s3-sqs input in that Fluentd S3 plugin- what does it exactly do and what's the use case?

**Comment by rverma-jm** [comment]
s3 has sqs/sns eventing inbuilt, for each added to s3 it push a message on sqs. One can listen to the message and read the logfile to publish the content to log forwarding destination. 
The alternate is discover new log file using s3 list object, which is very costly and will slow down as number of file increases.

**Comment by rverma-jm** [comment]
The use. case if being forwarding aws platform logs(WAF,DNS,Cloudtrail) directly to es from s3 destination. Although aws provide a way to deliver them using cloudwatch->kinesis->es, but that is too much overkill and costly.

**Comment by PettitWesley** [comment]
@rverma-jm interesting- I've never thought of that- using Fluent Bit to ship data between destinations in AWS. I've always thought of fluent bit as a collector you use to get data into AWS.

I've heard of people using Lambda for that- trigger a lambda on S3 upload which copies the object to your new destination.

Right now, I am going to prioritize getting a core Fluent Bit output plugin for S3 developed (I have a prototype already)- hopefully for release in a few months. The input is something we'll have to think more about.

I want to get your thoughts on the S3 output. Right now, the prototype has a significant limitation- it creates lots of little files in S3. 

The way Fluent Bit output plugins work is that Fluent Bit sends them data in chunks of a few thousand records. The plugin is supposed to handle those records, send them to the destination, and tell Fluent Bit whether it was successful or whether a retry is needed. Output plugins are not supposed to do their own buffering.

This means that the S3 output will create a file in S3 for every few thousand log lines. So each file in S3 will be very small, and there will be lots of them. In order for it to create larger files, Fluent Bit's buffering needs to be more configurable.

@rverma-jm Do you think that limitation is acceptable and that the plugin would be usable? I want to release an S3 output and then afterwards work on making buffering configurable.

**Comment by rverma-jm** [comment]
@PettitWesley I understand your view point, but lambdas doesn't handle a typical buffering capabilities. Also transformations/enrichments are easier with fluentbit.

Regarding s3 output, honestly we don't use s3 output at all. The main limitation is we query logs using athena, for that we prefer log transformation to orc. Kinesis firehose has this functionality. On top of that we have sql/flink analytics engine integrated into that. 

The essential requirement for s3 output plugin would be
1. Create large buffers, do sort within the buffers (i.e best effort sort logs by timestamp collected through multiple replicasets per buffer.)
2. Convert the buffer to parquet/arrow/orc.

I understand this deviates from your effort, but I believe it would still be useful for the community. Thanks for the efforts.

**Comment by rverma-jm** [comment]
Usually I found nothing beats s3 cli to be one of best performant s3 uploader. Wondering if we can share some code behind that.

Suggesting some required tuning parameters based on that
```
max_concurrent_requests - The maximum number of concurrent requests.
max_queue_size - The maximum number of tasks in the task queue.
multipart_threshold - The size threshold the CLI uses for multipart transfers of individual files.
multipart_chunksize - When using multipart transfers, this is the chunk size that the CLI uses for multipart transfers of individual files.
max_bandwidth - The maximum bandwidth that will be consumed for uploading and downloading data to and from Amazon S3.
use_accelerate_endpoint - Use the Amazon S3 Accelerate endpoint for all s3 and s3api commands
```

**Comment by PettitWesley** [comment]
Planned for end of september/beginning of october launch: https://github.com/aws/containers-roadmap/issues/940

**Comment by Missshao** [comment]
I noticed fluentd support output S3, is the feature same as this?
https://docs.fluentd.org/output/s3

**Comment by PettitWesley** [comment]
@Missshao Yeah, though the configuration options and behavior might not quite be the same.

**Comment by PettitWesley** [comment]
Fluent Bit S3 output support has been released. I'm changing the request title to Fluent Bit S3 input, since that doesn't exist.

**Comment by tiagomeireles** [comment]
+1 for this feature.

We want to use S3 as a buffer. We would have the fluentbit clients upload to s3 and then have a central fluentbit stack consuming from s3 and forwarding it to various other backends(loki, es etc..).

**Comment by SerhiiSokolov** [comment]
Hi there.
Is there any progress? We'd like to have an option to send logs from S3 (LB, WAF, etc) to Loki.

**Comment by psyhomb** [comment]
Is there anything new on this?
It would be incredibly valuable to have S3 input plugin for flunet-bit, something similar to [this](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-s3.html).

Thanks

**Comment by PettitWesley** [comment]
At the moment AWS does not have any near term plans to build an S3 input. 

If anyone in the community would like to try to implement it, I can help.

**Comment by andrejvanderzee** [comment]
@PettitWesley Is this still the case? We also might be interested in replacing filebeat+`aws-s3` input by fluent-bit, but it seems fluent-bit does not support the `aws-s3`-alike input. Still nothing on the roadmap?

**Comment by mashail** [comment]
for our case to read the logs archived in s3 bucket to reprocessing them as we are indexing them in loki but we can not keep them for long in loki .
